ntp-udp/src/raw_socket.rs

```diff
@@ -230,23 +294,32 @@ mod recv_message {
             // Invariants ensure that self.mhdr points to a valid libc::msghdr with a valid control
             // message region, and that self.next_msg either points to a valid control message
             // or is NULL.
-            // Since the previous statement ensures self.next_msg is not null, both passed
+            // The previous statement would have returned if self.next_msg were NULL, therefore both passed
             // pointers are valid for use with CMSG_NXTHDR
             // Invariant preservation:
             // CMSG_NXTHDR returns either a pointer to the next valid control message in the control
             // message region described by self.mhdr, or NULL
             self.next_msg = unsafe { libc::CMSG_NXTHDR(&self.mhdr, self.next_msg) };
 
             Some(match (current_msg.cmsg_level, current_msg.cmsg_type) {
-                (libc::SOL_SOCKET, libc::SO_TIMESTAMPING) => {
+                #[cfg(target_os = "linux")]
+                (libc::SOL_SOCKET, libc::SO_TIMESTAMPING | libc::SO_TIMESTAMPNS) => {
                     // Safety:
-                    // current_msg was constructed from a pointer that pointed to a valid
-                    // control message.
-                    // SO_TIMESTAMPING always has a timespec in the data
+                    // current_msg was constructed from a pointer that pointed to a valid control message.
+                    // SO_TIMESTAMPING and SO_TIMESTAMPNS always have a timespec in the data
                     let cmsg_data =
                         unsafe { libc::CMSG_DATA(current_msg) } as *const libc::timespec;
                     let timespec = unsafe { std::ptr::read_unaligned(cmsg_data) };
-                    ControlMessage::Timestamping(timespec)
+                    ControlMessage::Timestamping(read_ntp_timestamp_timespec(timespec))
+                }
+
+                (libc::SOL_SOCKET, libc::SO_TIMESTAMP) => {
+                    // Safety:
+                    // current_msg was constructed from a pointer that pointed to a valid control message.
+                    // SO_TIMESTAMP always has a timeval in the data
+                    let cmsg_data = unsafe { libc::CMSG_DATA(current_msg) } as *const libc::timeval;
+                    let timeval = unsafe { std::ptr::read_unaligned(cmsg_data) };
+                    ControlMessage::Timestamping(read_ntp_timestamp_timeval(timeval))
```

Comment:
Bit of a design nit, but this now makes this function do 2 things instead of 1, namely:
 - Provide a safe interface to libc
 - Handle conversion into ntp types
I would suggest we stick with just the first here to keep the unsafe code simpler and handle the conversion a layer up. This will likely also make it easier once we want to start sharing code with ptp

Reply:
yeah. We have to deal with both `timespec` and `timeval` here, but I found a way to split it out of this function at least.